src/lib.rs

```diff
@@ -132,6 +137,140 @@ impl Invariant {
         Ok(())
     }
 
+    pub async fn create_position(
+        &mut self,
+        pool_key: PoolKey,
+        lower_tick: i32,
+        upper_tick: i32,
+        liquidity_delta: Liquidity,
+        slippage_limit_lower: SqrtPrice,
+        slippage_limit_upper: SqrtPrice,
+    ) -> Result<Position, InvariantError> {
+        let caller = msg::source();
+        let program = exec::program_id();
+        let current_timestamp = exec::block_timestamp();
+        let current_block_number = exec::block_height() as u64;
+
+        // liquidity delta = 0 => return
+        if liquidity_delta == Liquidity::new(0) {
+            return Err(InvariantError::ZeroLiquidity);
+        }
+
+        let mut pool = self.pools.get(&pool_key)?;
+
+        let mut lower_tick = self
+            .ticks
+            .get(pool_key, lower_tick)
+            .cloned()
+            .unwrap_or_else(|_| Self::create_tick(self, pool_key, lower_tick).unwrap());
+
+        let mut upper_tick = self
+            .ticks
+            .get(pool_key, upper_tick)
+            .cloned()
+            .unwrap_or_else(|_| Self::create_tick(self, pool_key, upper_tick).unwrap());
+
+        let (position, x, y) = Position::create(
+            &mut pool,
+            pool_key,
+            &mut lower_tick,
+            &mut upper_tick,
+            current_timestamp,
+            liquidity_delta,
+            slippage_limit_lower,
+            slippage_limit_upper,
+            current_block_number,
+            pool_key.fee_tier.tick_spacing,
+        )?;
+
+        self.pools.update(&pool_key, &pool)?;
+
+        self.positions.add(&caller, &position);
+
+        self.ticks.update(pool_key, lower_tick.index, lower_tick)?;
+        self.ticks.update(pool_key, upper_tick.index, upper_tick)?;
+
+        self.transfer_tokens(&pool_key.token_x, None, &caller, &program, x.get()).await?;
```

Comment:
Can we test the consistency of this entrypoint in specific e2e tests? For instance, if the user is missing "token_y" and there is an insufficient amount of "token_x"